jxl/src/frame/modular/tree.rs

```diff
@@ -167,4 +177,202 @@ impl Tree {
             histograms,
         })
     }
+
+    pub fn max_property(&self) -> usize {
+        self.nodes
+            .iter()
+            .map(|x| match x {
+                TreeNode::Leaf { .. } => 0,
+                TreeNode::Split { property, .. } => *property,
+            })
+            .max()
+            .unwrap() as usize
+    }
+
+    pub fn num_prev_channels(&self) -> usize {
+        self.max_property().saturating_sub(NUM_NONREF_PROPERTIES) / PROPERTIES_PER_PREVCHAN
+    }
+
+    // Note: `property_buffer` is passed as input because this implementation relies on having the
+    // previous values available for computing the local gradient property.
+    // Also, the first two properties (the static properties) should be already set by the caller.
+    // All other properties should be 0 on the first call in a row.
+    #[allow(clippy::too_many_arguments)]
+    #[instrument(level = "trace", skip(buffers), ret)]
+    pub(super) fn predict(
+        &self,
+        buffers: &mut [ModularBufferInfo],
+        buffer_indices: &[usize],
+        index: usize,
+        grid_index: usize,
+        wp_state: &mut WeightedPredictorState,
+        x: usize,
+        y: usize,
+        property_buffer: &mut [i32; 256],
+    ) -> PredictionResult {
+        let get_pixel = |x: usize, y: usize| -> i32 {
```

Comment:
I think one could make the code below even more readable, if one that function was name `get_pixel_or_default` and a `condition` and `default` argument. Then one could save the last else {} in every case below, no?
e.g. the first calls would then be 
```
left = get_pixel_or_default(x - 1, y, x > 0, get_pixel_or_default(x, y - 1, y > 0, 0)
top = get_pixel_or_default(x, y - 1, y > 0, left)
```

Reply:
..it is also fine as is

Reply:
not sure it would be more readable...